webdriver/support/fixtures.py

```diff
@@ -80,30 +81,49 @@ def http(session):
 def server_config():
     return json.loads(os.environ.get("WD_SERVER_CONFIG"))
 
-def session(session_session, request):
-    # finalisers are popped off a stack,
-    # making their ordering reverse
-    request.addfinalizer(lambda: _switch_to_top_level_browsing_context(session_session))
-    request.addfinalizer(lambda: _restore_windows(session_session))
-    request.addfinalizer(lambda: _dismiss_user_prompts(session_session))
-
-    return session_session
-
-# Create a wdclient `Session` object for each Pytest "session"
-def session_session(request):
-    host = os.environ.get("WD_HOST", default_host)
-    port = int(os.environ.get("WD_PORT", default_port))
-    capabilities = json.loads(os.environ.get("WD_CAPABILITIES", "{}"))
-
-    session = webdriver.Session(host, port, desired_capabilities=capabilities)
-
-    def destroy():
-        if session.session_id is not None:
-            session.end()
-
-    request.addfinalizer(destroy)
-
-    return session
+# Provide a factory function that produces wdclient `Session` instances. If the
+# `WD_CAPABILITIES` environment variable is set, it will be parsed as JSON and
+# the resulting object will be included in the WebDriver "Create Session"
+# command. Additional capabilities may be specified as an optional argument to
+# this function, but the operation will fail if any values conflict with those
+# specified via the environment. If the session is still active at the
+# completion of the test, it will be destroyed automatically.
+def create_session(request):
+    def create_session(test_capabilities=None):
+        host = os.environ.get("WD_HOST", default_host)
+        port = int(os.environ.get("WD_PORT", default_port))
+        if test_capabilities is None:
+            test_capabilities = {}
+        env_capabilities = json.loads(os.environ.get("WD_CAPABILITIES", "{}"))
+
+        always_match = merge_dictionaries(env_capabilities.get("always_match", {}),
+                                          test_capabilities.get("always_match", {}))
+        first_match = merge_dictionaries(env_capabilities.get("first_match", {}),
+                                         test_capabilities.get("first_match", {}))
```

Comment:
Yes, for sure. (The boring background is that this patch started out with the
two objects defined as independent variables, which informed my choice of
"snake cased" names.)